src/key/key.js

```diff
@@ -360,45 +360,62 @@ class Key {
       throw new Error('Primary key is expired');
     }
+    // check for expiration time in direct signatures
+    const directSignature = await helper.getLatestValidSignature(
+      this.directSignatures, primaryKey, enums.signature.key, { key: primaryKey }, date, config
+    ).catch(() => {}); // invalid signatures are discarded, to avoid breaking the key
+
+    if (directSignature && helper.isDataExpired(primaryKey, directSignature, date)) {
+      throw new Error('Primary key is expired');
+    }
   }
 
   /**
-   * Returns the latest date when the key can be used for encrypting, signing, or both, depending on the `capabilities` paramater.
-   * When `capabilities` is null, defaults to returning the expiry date of the primary key.
-   * Returns null if `capabilities` is passed and the key does not have the specified capabilities or is revoked or invalid.
-   * Returns Infinity if the key doesn't expire.
-   * @param  {encrypt|sign|encrypt_sign} [capabilities] - capabilities to look up
-   * @param  {module:type/keyid~KeyID} [keyID] - key ID of the specific key to check
+   * Returns the expiration date of the key.
+   * If no keyID is specified, it returns the expiration date of the primary key.
+   * Returns `Infinity` if the key doesn't expire, or `null` if the key is revoked or invalid.
+   * @param  {module:type/keyid~KeyID} [keyID] - key ID of the specific (sub)key to check
    * @param  {Object} [userID] - User ID to consider instead of the primary user
    * @param  {Object} [config] - Full configuration, defaults to openpgp.config
    * @returns {Promise<Date | Infinity | null>}
    * @async
    */
-  async getExpirationTime(capabilities, keyID, userID, config = defaultConfig) {
-    const primaryUser = await this.getPrimaryUser(null, userID, config);
-    const selfCert = primaryUser.selfCertification;
-    const keyExpiry = helper.getKeyExpirationTime(this.keyPacket, selfCert);
-    const sigExpiry = selfCert.getExpirationTime();
-    let expiry = keyExpiry < sigExpiry ? keyExpiry : sigExpiry;
-    if (capabilities === 'encrypt' || capabilities === 'encrypt_sign') {
-      const encryptKey =
-        await this.getEncryptionKey(keyID, expiry, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {}) ||
-        await this.getEncryptionKey(keyID, null, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {});
-      if (!encryptKey) return null;
-      const encryptExpiry = await encryptKey.getExpirationTime(null, config);
-      if (encryptExpiry < expiry) expiry = encryptExpiry;
+  async getExpirationTime(keyID, userID, config = defaultConfig) {
+    if (keyID) {
+      const keys = this.getKeys(keyID);
+      if (keys.length === 0) {
+        throw new Error(`Could not find key with key ID ${keyID.toHex()}`);
+      }
+      const [key] = keys;
+      if (key instanceof Subkey) {
+        return key.getExpirationTime(null, config);
+      }
     }
```

Comment:
Let's just get rid of this parameter too, callers can do this manually if this is what they want